satpy/writers/cf_writer.py

```diff
@@ -160,16 +164,121 @@ def make_time_bounds(dataarray, start_times, end_times):
                         dims=['time_bnds'], coords={'time_bnds': [0, 1]})
 
 
+class AttributeEncoder(json.JSONEncoder):
+    """JSON encoder for dataset attributes"""
+    def default(self, obj):
+        """Returns a json-serializable object for 'obj'
+
+        In order to facilitate decoding, elements in dictionaries, lists/tuples and multi-dimensional arrays are
+        encoded recursively.
+        """
+        if isinstance(obj, dict):
+            serialized = {}
+            for key, val in obj.items():
+                serialized[key] = self.default(val)
+            return serialized
+        elif isinstance(obj, (list, tuple, np.ndarray)):
+            return [self.default(item) for item in obj]
+        return self._encode(obj)
+
+    def _encode(self, obj):
+        """Encode the given object as a json-serializable datatype.
+
+        Use the netcdf encoder as it covers most of the datatypes appearing in dataset attributes. If that fails,
+        return the string representation of the object."""
+        try:
+            return _encode_nc(obj)
+        except ValueError:
+            return str(obj)
+
+
+def _encode_nc(obj):
+    """Encode an arbitrary object in a netcdf compatible datatype
+
+    Raises:
+        ValueError if no netcdf compatible datatype could be found
+    """
+    if isinstance(obj, (bool, np.bool_)):
+        # Bool has to be checked first, because it is a subclass of int
+        return str(obj)
+    elif isinstance(obj, (int, float, str)):
+        return obj
+    elif isinstance(obj, np.integer):
+        return int(obj)
+    elif isinstance(obj, np.floating):
+        return float(obj)
+    elif isinstance(obj, np.void):
+        return tuple(obj)
+    elif isinstance(obj, np.ndarray):
+        if not len(obj.dtype) and obj.dtype == np.bool_:
```

Comment:
Regarding the CodeFactor issue: I'm using `len(obj.dtype)` to identify record arrays